WalletWasabi/Blockchain/Keys/KeyManager.cs

```diff
@@ -429,68 +380,32 @@ public void SetDoNotSkipSynchronization()
 			({ } k, { } i) => GetKeys(x => x.IsInternal == i && x.KeyState == k)
 		};
 
-	/// <param name="ignoreTail">If true it does only consider the gap between used keys and does not care about the nonused keys at the end.</param>
-	public int CountConsecutiveUnusedKeys(bool isInternal, bool ignoreTail)
+	public int CountConsecutiveUnusedKeys(bool isInternal, ScriptPubKeyType scriptPubKeyType = ScriptPubKeyType.Segwit)
 	{
-		var keyIndexes = GetKeys(x => x.IsInternal == isInternal && x.KeyState != KeyState.Used).Select(x => x.Index).ToArray();
-
-		if (ignoreTail)
-		{
-			var lastUsedIndex = GetKeys(x => x.IsInternal == isInternal && x.KeyState == KeyState.Used).LastOrDefault()?.Index;
-			if (lastUsedIndex is null)
-			{
-				return 0;
-			}
-			else
-			{
-				keyIndexes = keyIndexes.Where(x => x < lastUsedIndex).ToArray();
-			}
-		}
-
-		var hs = keyIndexes.ToHashSet();
-		int largerConsecutiveSequence = 0;
-
-		for (int i = 0; i < keyIndexes.Length; ++i)
-		{
-			if (!hs.Contains(keyIndexes[i] - 1))
-			{
-				int j = keyIndexes[i];
-				while (hs.Contains(j))
-				{
-					j++;
-				}
-
-				var sequenceLength = j - keyIndexes[i];
-				if (largerConsecutiveSequence < sequenceLength)
-				{
-					largerConsecutiveSequence = sequenceLength;
-				}
-			}
-		}
-		return largerConsecutiveSequence;
+		var keySource = GetHdPubKeyGenerator(isInternal, scriptPubKeyType);
+		var view = HdPubKeyCache.GetView(keySource.KeyPath);
```

Comment:
It doesn't matter. Basically this method should be removed completely. I didn't do it (yet) because I would need to modify the UI. It is used to display the gap (nonsense) and it not used for anything useful. 

Technically speaking there exists the really tiny (i would say that unfeasible) chance that the user could get a `Collection was modified......` exception but again, it is only in theory and we have to remove  this thing so users will never see it.

Reply:
Btw, and just as a note: it used to be used/accessed for something critically important without a lock!